heartbeat/gcp-vpc-move-route.in

```diff
@@ -0,0 +1,441 @@
+  check_conflicting_routes
+
+  try:
+    request = ctx.conn.routes().get(project=ctx.project, route=ctx.route_name)
+    response = request.execute()
+  except googleapiclient.errors.HttpError as e:
+    if 'Insufficient Permission' in e.content:
+      return OCF_ERR_PERM
+    elif e.resp.status == 404:
+      return OCF_NOT_RUNNING
+    else:
+      raise
+
+  routed_to_instance = response.get('nextHopInstance', '<unknown>')
+  instance_url = '%s/projects/%s/zones/%s/instances/%s' % (
+      GCP_API_URL_PREFIX, ctx.project, ctx.zone, ctx.instance)
+  if routed_to_instance != instance_url:
+    logging.warn(
+        'The floating IP %s is not routed to this instance (%s) but to instance %s',
+        ctx.ip, ctx.instance, routed_to_instance.split('/')[-1])
+    return OCF_NOT_RUNNING
+
+  logging.debug(
+      'The floating IP %s is correctly routed to this instance (%s)',
+      ctx.ip, ctx.instance)
+  return OCF_SUCCESS
+
+
+def ip_and_route_stop(ctx):
+  logging.info('Bringing down the floating IP %s', ctx.ip)
+
+  # Delete the route entry
+  # If the route entry exists and is pointing to another instance, don't touch it
+  if route_monitor(ctx) == OCF_NOT_RUNNING:
+    logging.info(
+        'The floating IP %s is already not routed to this instance (%s)',
+        ctx.ip, ctx.instance)
+  else:
+    route_release(ctx)
+
+  if ip_monitor(ctx) == OCF_NOT_RUNNING:
+    logging.info('The floating IP %s is already down', ctx.ip)
+  else:
+    ip_release(ctx)
+
+
+def configure_logs(ctx):
+  # Prepare logging
+  logging.basicConfig(
+      format='gcp:route - %(levelname)s - %(message)s', level=logging.INFO)
+  logging.getLogger('googleapiclient').setLevel(logging.WARN)
+  logging_env = os.environ.get('OCF_RESKEY_stackdriver_logging')
+  if logging_env:
+    logging_env = logging_env.lower()
+    if any(x in logging_env for x in ['yes', 'true', 'enabled']):
+      try:
+        import google.cloud.logging.handlers
+        client = google.cloud.logging.Client()
+        handler = google.cloud.logging.handlers.CloudLoggingHandler(
+            client, name=ctx.instance)
+        handler.setLevel(logging.INFO)
+        formatter = logging.Formatter('gcp:route "%(message)s"')
+        handler.setFormatter(formatter)
+        root_logger = logging.getLogger()
+        root_logger.addHandler(handler)
+      except ImportError:
+        logging.error('Couldn\'t import google.cloud.logging, '
+            'disabling Stackdriver-logging support')
+
+
+def main():
+  if 'meta-data' in sys.argv[1]:
+    print(METADATA)
+    return
+
+  ctx = Context()
+
+  validate(ctx)
+  if 'validate-all' in sys.argv[1]:
+    return
+
+  configure_logs(ctx)
+  if 'start' in sys.argv[1]:
+    ip_and_route_start(ctx)
+  elif 'stop' in sys.argv[1]:
+    ip_and_route_stop(ctx)
+  elif 'status' in sys.argv[1] or 'monitor' in sys.argv[1]:
+    sys.exit(ip_monitor(ctx))
+  else:
+    usage = 'usage: $0 {start|stop|monitor|status|meta-data|validate-all}'
```

Comment:
$0 is os.path.basename(sys.argv[0]) in Python.

Reply:
fixed